### PR TITLE
fix: use single connection to sync all mailboxes

### PR DESCRIPTION
**Pull request by SebastianKrupinski**
Resolves: https://github.com/nextcloud/mail/issues/10098

### CODE REVIEW

File: lib/Service/Sync/ImapToDbSynchronizer.php [hunk 1]
```diff
@@ -109,6 +109,10 @@ public function syncAccount(Account $account,
 		$snoozeMailboxId = $account->getMailAccount()->getSnoozeMailboxId();
 		$sentMailboxId = $account->getMailAccount()->getSentMailboxId();
 		$trashRetentionDays = $account->getMailAccount()->getTrashRetentionDays();
+		// construct imap client and perform initial connection
```
Comment (kesselb) [inline lib/Service/Sync/ImapToDbSynchronizer.php:112]:
```suggestion
```

https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/#h2-b7593c698a940

File: lib/Service/Sync/ImapToDbSynchronizer.php [hunk 2]
```diff
@@ -109,6 +109,10 @@ public function syncAccount(Account $account,
...
+		$client = $this->clientFactory->getClient($account);
+		$client->login();
+		// perform mail sync on all mail boxes
```
Comment (kesselb) [inline lib/Service/Sync/ImapToDbSynchronizer.php:115]:
```suggestion
```

https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/#h2-b7593c698a940

File: lib/Service/Sync/ImapToDbSynchronizer.php [hunk 3]
```diff
@@ -131,6 +136,8 @@ public function syncAccount(Account $account,
 				$rebuildThreads = true;
 			}
 		}
+		// disconnect imap client
```
Comment (kesselb) [inline lib/Service/Sync/ImapToDbSynchronizer.php:139]:
```suggestion
```
https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/#h2-b7593c698a940

File: lib/Service/Sync/SyncService.php [hunk 4]
```diff
@@ -29,6 +30,9 @@
 use function array_map;
 
 class SyncService {
+	/** @var IMAPClientFactory */
+	private $clientFactory;
```
Comment (kesselb) [inline lib/Service/Sync/SyncService.php:34]:
```suggestion
	private IMAPClientFactory $clientFactory;
```

File: lib/Service/Sync/SyncService.php [hunk 5]
```diff
@@ -103,9 +110,13 @@ public function syncMailbox(Account $account,
 		if ($partialOnly && !$mailbox->isCached()) {
 			throw MailboxNotCachedException::from($mailbox);
 		}
-
+		// construct imap client and perform initial connection
```
Comment (kesselb) [inline lib/Service/Sync/SyncService.php:113]:
```suggestion
```

https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/#h2-b7593c698a940

File: lib/Service/Sync/SyncService.php [hunk 6]
```diff
@@ -103,9 +110,13 @@ public function syncMailbox(Account $account,
...
+		$client = $this->clientFactory->getClient($account);
+		$client->login();
+		// perform mail sync on all mail boxes
```
Comment (kesselb) [inline lib/Service/Sync/SyncService.php:116]:
```suggestion
```

https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/#h2-b7593c698a940

File: lib/Service/Sync/SyncService.php [hunk 7]
```diff
@@ -114,6 +125,8 @@ public function syncMailbox(Account $account,
 		);
 
 		$this->mailboxSync->syncStats($account, $mailbox);
+		// disconnect imap client
```
Comment (kesselb) [inline lib/Service/Sync/SyncService.php:128]:
```suggestion
```

https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/#h2-b7593c698a940

File: tests/Unit/Service/Sync/SyncServiceTest.php [hunk 8]
```diff
@@ -25,6 +27,13 @@
 use Psr\Log\LoggerInterface;
 
 class SyncServiceTest extends TestCase {
+
+	/** @var IMAPClientFactory */
+	private $clientFactory;
+
+	/** @var Horde_Imap_Client_Socket */
+	private $client;
```
Comment (kesselb) [inline tests/Unit/Service/Sync/SyncServiceTest.php:35]:
```suggestion
	private IMAPClientFactory&MockObject $clientFactory;
	private Horde_Imap_Client_Socket&MockObject $client;
```

and import `use PHPUnit\Framework\MockObject\MockObject;`

**Review by kesselb — CHANGES_REQUESTED** [review]

File: lib/Service/Sync/ImapToDbSynchronizer.php [hunk 9]
```diff
@@ -109,6 +109,10 @@ public function syncAccount(Account $account,
 		$snoozeMailboxId = $account->getMailAccount()->getSnoozeMailboxId();
 		$sentMailboxId = $account->getMailAccount()->getSentMailboxId();
 		$trashRetentionDays = $account->getMailAccount()->getTrashRetentionDays();
+		// construct imap client and perform initial connection
+		$client = $this->clientFactory->getClient($account);
+		$client->login();
```
Comment (ChristophWurst) [inline lib/Service/Sync/ImapToDbSynchronizer.php:114]: This …

File: lib/Service/Sync/ImapToDbSynchronizer.php [hunk 10]
```diff
@@ -202,10 +210,6 @@ public function sync(Account $account,
 		if ($mailbox->getSelectable() === false) {
 			return $rebuildThreads;
 		}
-
-		$client = $this->clientFactory->getClient($account);
-		$client->login(); // Need to login before fetching capabilities.
```
Comment (ChristophWurst) [inline lib/Service/Sync/ImapToDbSynchronizer.php:207]: … should go back here. The login outside seems arbitrary, so the next developer might drop it.

1. Move the login back to where it matters
2. Do not drop the comment explaining why we need the explicit login

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Review by st3iny — APPROVED** [review]

**Pull request merged**